### DISCUSSION TITLE
Double counting a statement termination token

### DISCUSSION THREAD

**Discussion by Pyromuffin** [tree-sitter/tree-sitter#988] [Q&A] [answered]
Hi, I have an interesting problem.
I am writing a parser for the inform7 language, which is modeled after natural language.

You end most statements with a "." for example:
`A dog is in the room.`
In this case, the statement termination is ".", a full stop.

However, like with natural language, you can end a statement with the "." inside of a string. for example:
`The description of the dog is "a good boy."`

In this case the termination token is... a string? It's easy enough to create a symbol for a string containing statement termination.
The problem I am having is if I make a $.terminated_string statement termination, then it would not be considered an expression for the purpose of the right hand side of the "is" statement.

It would seem I need to double count  or emit two tokens for this terminated string as both an expression and as statement termination. I am not sure how to do that.

Thanks,
Kelly

edit: 
Even trying this with an external scanner, I'm not sure how this would work without the ability to look behind.

**Comment by tek** [comment]
what would be the intended parse tree of the examples?

**Reply by Pyromuffin** [reply]
`The description is "a good boy".`

```
(simple_statements [3, 0] - [3, 32]
    (simple_statement [3, 0] - [3, 31]
      (expression_statement [3, 0] - [3, 31]
        (is_fragment [3, 0] - [3, 31]
          (identifier [3, 0] - [3, 16]
            (definite_article [3, 0] - [3, 4])
            (word [3, 4] - [3, 15]))
          (string_literal [3, 19] - [3, 31]))))
    (fragment_termination [3, 31] - [3, 32])))
```

I would like `The description is "a good boy."` to parse the same way, basically.
I've taken the basic structure of the statement grammar, and external scanner from the python grammar, because blocks are delimited by levels indentation. This means that the `fragment_termination` is not a part of each statement's pattern, but a part of `simple statements` which can either end in a newline, or a termination token (which I have called fragment termination).

**Reply by tek** [reply]
What is your intention for having `fragment_termination` as a node like this? Do you have a use case?

You could have that node inside of `string_literal_with_terminator`, for example. Could that fulfill the same purpose?

**Reply by Pyromuffin** [reply]
Here is the grammar for simple statements (which is copied directly from the python grammar, with the name for fragment_termination changed):

```
     simple_statements: $ => seq(
            $.simple_statement,
            optional(repeat(seq(
              $.fragment_termination,
              $.simple_statement
            ))),
            optional($.fragment_termination),
            $._newline
          ),
```

the idea behind this is that you can place several statements on the same line as long as they are delimited by fragment termination. example:
```
A dog is in the room. The description is "a good boy". The dog carries a toy. 
```
Which works fine. Changing the fragment termination to be part of the string node would make it so that it wouldn't be part of the simple statements node, which would make the following fail to parse because it would not find fragment termination where it expects it. (unless there is something I am missing?)
```
A dog is in the room. The description is "a good boy." The dog carries a toy. 
```

edit:
replacing $.fragment_termination with choice($.fragement_termination, $.string_literal_with_terminator) wouldn't work either because then it wouldn't match the RHS of the is_fragment. 
I think the core of the problem is that $.string_literal_with_terminator needs to be both an expression and a terminator.

**Comment by tek** [comment]
my suggestion would be something along these lines:
```javascript
statement: $ => seq(
  repeat1(choice($.word, $.literal)),
  choice($.termination, $.literal_with_termination),
)
```

You'd need a conflict for the literals, I'd assume, but the parser would handle the disambiguation for you.
It's important, of course, that you don't allow the `.` inside the `literal` rule.

**Reply by Pyromuffin** [reply]
yeah that would work if there was only one kind of statement. There are a lot of kinds of statements though, so I would have to duplicate each simple statement type.

another method i'm consdiering:  i might be able to restructure the way string literals are parsed, using an external scanner. I could have:
an external token for '"' (double quote) that starts a string, set a flag in the external scanner to know we're inside a string literal. 
then, i can have a string literal end in either '"' or /. *"/, and that second token could be considered fragment termination, and wouldn't have to be a child of an expression.

Haven't tried it yet, and it seems annoying to figure out the size of a string literal this way for syntax highlighting purposes, but it might work.

**Reply by tek** [reply]
do you have the grammar online?

**Reply by ahlinc** [reply]
@tek https://github.com/Pyromuffin/tree-sitter-inform7/blob/main/grammar.js

**Reply by ahlinc** [reply, marked as answer]
@tek I saw that in the mentioned grammar there are a lot of another issues that it seems affect mentioned problem.
1. There is a mistake approach how to spaces are processed. You can see on the bellow screenshot that spaces are consumed into some tokens.
2. There issues with simple phrases:

![image](https://user-images.githubusercontent.com/14666676/111163053-6ffece00-85a5-11eb-8a29-e6da2512a107.png)


Helper script: 
```bash
i7-parse() {
    {
        echo '"Foobar" by No One'
        echo
        echo -n "$1"
    } | \
    tree-sitter parse "${@:2}" -c /dev/stdin
}
```

The `-c` parameter is from there #993 

Also I've made some cleanups there https://github.com/ahlinc/tree-sitter-inform7 in the `main` branch and a PR https://github.com/Pyromuffin/tree-sitter-inform7/pull/1

**Reply by tek** [reply]
@ahlinc did you add the annotations with the parsed text manually?

**Reply by ahlinc** [reply]
@tek Of course not :smile: See this PR: #993
It can't show text comments for all tokens yet.

**Reply by Pyromuffin** [reply]
@ahlinc, thanks for the cleanup. Yes, the way whitespace is processed is kind of weird, because whitespace is required to come after some tokens, but it is unrelated to the problem with `The dog carries a toy.` If there is a better way to handle whitespace, please feel free to let me know.

The reason that fails to parse is that I haven't pushed support for yet. Unfortunately, there's a bit of a problem with parsing "carries" as a verb, or other verb like sentences that can be generated in the source text, so in general, the semantics of that sentence needs to be handled after parsing. At any rate, it should just be parsed as an identifier expression statement.

**Reply by tek** [reply]
> @tek Of course not smile See this PR: #993
> It can't show text comments for all tokens yet.

oohhh, very nice!

**Comment by tek** [comment]
I can't make an informed judgement yet, but it does seem that you don't need to rely on the terminating token everywhere to decide that the statement is terminated.

In your example, it seems that string literals both with and without inside dot cannot be followed by another symbol legally, so using something like `choice(seq($.expression, $.fragment_termination), $.terminated_string))` appears plausible (you can use a hidden (inline) rule for this construct). I would assume that the left expression of an `is_fragment` cannot be terminated by a dot, so there it wouldn't be legal anyway.
An identifier on the other hand, as I understand it, can _not_ contain a string literal, so it is terminated differently (simpler).

So in general I would suggest that you identify and separate the statements that can be terminated with the simple method and add the termination token there as a following token and handle the special cases individually.
